linera-chain/src/chain.rs

```diff
@@ -1180,43 +1203,83 @@ where
                 *outbox_counters.entry(height).or_default() += 1;
             }
         }
-        let full_names = raw_outcome
-            .subscribe
+
+        self.subscribe(raw_outcome.subscribe, application_id, false)
+            .await?;
+        Ok(())
+    }
+
+    async fn subscribe(
+        &mut self,
+        names_and_ids: Vec<(ChannelName, ChainId)>,
+        application_id: GenericApplicationId,
+        send_all: bool,
+    ) -> Result<(), ChainError> {
+        let full_names = names_and_ids
             .iter()
-            .map(|(name, _id)| ChannelFullName {
+            .map(|(name, _)| ChannelFullName {
                 application_id,
                 name: name.clone(),
             })
             .collect::<Vec<_>>();
         let channels = self.channels.try_load_entries_mut(&full_names).await?;
-        let subscribe_channels = raw_outcome.subscribe.into_iter().zip(channels);
+        let subscribe_channels = names_and_ids.into_iter().zip(channels);
+        let max_stream_queries = self.context().max_stream_queries();
         let stream = stream::iter(subscribe_channels)
             .map(|((name, id), mut channel)| async move {
-                let mut result = None;
+                if channel.subscribers.contains(&id).await? {
+                    return Ok(None); // Was already a subscriber.
+                }
+                channel.subscribers.insert(&id)?;
+                let heights = if send_all {
+                    // Send all messages.
```

Comment:
No, `outbox.schedule_message` below filters out the messages that have already been sent. (I don't think we actually ever delete the outbox?)

We could already filter here, of course, but that wouldn't really make a difference: We'd just load the same outboxes a bit earlier, and do the filtering twice, effectively.

Reply:
We could consider using a `CustomSetView`, ordered by descending height, and then [`for_each_index_while`](https://docs.rs/linera-views/latest/linera_views/set_view/struct.CustomSetView.html#method.for_each_index_while) until we reach a height below what's expected by the outbox. But I'm not sure that's actually faster, because it would read the heights sequentially then.